### PR TITLE
Prefix CMake options with RAJA_

### PR DESCRIPTION
**Pull request by davidbeckingsale**
- This PR is a refactoring.
- Adds the RAJA_ prefix to all CMake options.
- Options that shadow a BLT option are turned into cmake_dependent_option calls, ensuring that they can be controlled independently and have the correct dependence on the underlying BLT support (e.g. RAJA_ENABLE_CUDA requires ENABLE_CUDA)

Dependent options can be set by using the parent option, but other options will require modification of build scripts and will impact users.

### CODE REVIEW

File: RELEASE_NOTES.md
Comment (kab163) [inline RELEASE_NOTES.md:90]: How does this sound?

File: .travis.yml
Comment (kab163) [inline .travis.yml:37]: Should the "-fmodules" flag be here?

Comment (davidbeckingsale) [reply .travis.yml:37]: I think we can keep this if it's working.

File: cmake/SetupCompilers.cmake
Comment (kab163) [inline cmake/SetupCompilers.cmake:14]: Added this in the cmake to say if the RAJA_ENABLE_MODULES flag is used, print out a warning that it's deprecated and force it to be turned off instead - hope that's ok?

Comment (davidbeckingsale) [reply cmake/SetupCompilers.cmake:14]: Yes. I think you could add to the message that the flag `-fmodules` can be added manually if desired.

Comment (kab163) [reply cmake/SetupCompilers.cmake]: Sounds good!

File: RELEASE_NOTES.md
Comment (rhornung67) [reply RELEASE_NOTES.md:90]: Maybe replace "since it hasn't been kept up to date" with "since it can cause issues". Also, change "has been removed" to " will be removed in the next release". When something gets deprecated, we typically make users aware that it will be removed in the future.

Comment (kab163) [reply RELEASE_NOTES.md:90]: Sounds good! Thanks for edits

Comment (davidbeckingsale) [reply RELEASE_NOTES.md:90]: I just wanted to clarify one thing I think got lost in the email thread - the ENABLE_MODULES option has been broken for a long time and doesn't enable modules. The issue was that the code block to enable the flag was looking for the `RAJA_ENABLE_MODULES` variable, which didn't exist. This PR renamed the ENABLE_MODULES option and caused this block to be executed, where it didn't used to be.

So there's not really anything to "remove" in terms of the option because it never worked.

File: RELEASE_NOTES.md
Comment (rhornung67) [inline RELEASE_NOTES.md:89]: 👍

File: RELEASE_NOTES.md
Comment (rhornung67) [reply RELEASE_NOTES.md]: So @davidbeckingsale, you're advocating for just ripping it out? I guess that makes sense since if it didn't work, folks weren't using it, right?

Comment (davidbeckingsale) [reply RELEASE_NOTES.md]: Yeah. The only way to enable it is to manually add the `-fmodules` flag. The deprecation will come from removing the modulemap file which we can do next release

Comment (rhornung67) [reply RELEASE_NOTES.md]: @kab163 will you please add an issue to document in the user guide how to enable the modules for CMake and regular make based build systems? We will add for the next release. Thank you.

Comment (kab163) [reply RELEASE_NOTES.md]: Sure thing. I'll update this deprecation message too

File: cmake/SetupCompilers.cmake
Comment (davidbeckingsale) [inline cmake/SetupCompilers.cmake:13]:
```suggestion
  message(WARNING "RAJA_ENABLE_MODULES is deprecated, please add the -fmodules flag manually if desired.")
```

Comment (kab163) [reply cmake/SetupCompilers.cmake:13]: Ah yes, much more concise.

**Comment by kab163** [comment]
I think this is ready to merge!

File: docs/sphinx/user_guide/config_options.rst
Comment (rhornung67) [inline docs/sphinx/user_guide/config_options.rst:104]: I may be wrong, but think these changes are going to break the sphinx docs formatting. As I understand it, the '=' row delimeters need to extend at least as far as the test in each row. Adding more '=' so the columns align should fix it. Have you looked at the regenerated docs with these changes?

**Review by rhornung67 — CHANGES_REQUESTED** [review]
A question about this? I thought we would rely on CMake or BLT variable names for things like 'ENABLE_WARNINGS', etc. and set RAJA_ENABLE_WARNINGS based on those. Doesn't that make it easier when building multiple integrated packages?

**Comment by rhornung67** [comment]
This PR also breaks the doxygen build on RTD. I removed the errant addition of RAJA_ prefix to doxygen options in the Doxyfile. Unfortunately, the doxygen build is still broken. Digging into it...

File: docs/sphinx/user_guide/config_options.rst
Comment (rhornung67) [reply docs/sphinx/user_guide/config_options.rst:104]: Indeed, running text into the column margins breaks the sphinx build. I pushed changes to fix.

Comment (kab163) [reply docs/sphinx/user_guide/config_options.rst:104]: Thanks Rich!

**Comment by rhornung67** [comment]
@davidbeckingsale @kab163 the last changes I pushed to this branch fixed the sphinx and doxygen docs on RTD. When this is merged into develop, the latest docs should rebuild to fix the breakage there.

**Comment by rhornung67** [comment]
> A question about this? I thought we would rely on CMake or BLT variable names for things like 'ENABLE_WARNINGS', etc. and set RAJA_ENABLE_WARNINGS based on those. Doesn't that make it easier when building multiple integrated packages?

@davidbeckingsale please answer this question. i am a bit confused about how the RAJA_ cmake options work with the cmake or blt variants without the prefix.

**Comment by davidbeckingsale** [comment]
> > A question about this? I thought we would rely on CMake or BLT variable names for things like 'ENABLE_WARNINGS', etc. and set RAJA_ENABLE_WARNINGS based on those. Doesn't that make it easier when building multiple integrated packages?
> 
> @davidbeckingsale please answer this question. i am a bit confused about how the RAJA_ cmake options work with the cmake or blt variants without the prefix.

Yes, that's how it works (I think changing all the options in the docs makes it confusing) RAJA_ENABLE_WARNINGS will basically inherit the value from ENABLE_WARNINGS, but you can also manually set it to be off.

**Comment by rhornung67** [comment]
The prefix versions give flexibility to turn things on or off for only RAJA, correct?

We should add something to the docs about that. I can do that, but I'm not sure which options are 'duplicates' with cmake and blt. Do you know which ones are specific to RAJA?

**Comment by davidbeckingsale** [comment]
> The prefix versions give flexibility to turn things on or off for only RAJA, correct?
> 
> We should add something to the docs about that. I can do that, but I'm not sure which options are 'duplicates' with cmake and blt. Do you know which ones are specific to RAJA?

Yes, that's correct. You can look at cmake/SetupRajaOptions.cmake and cmake/SetupDependentOptions.cmake to see the two different types. SetupRajaOptions has the ones specific to RAJA.

**Comment by rhornung67** [comment]
Thanks. That helps. I'll work some changes into the sphinx docs. When it's pushed, I'll ping you to look it over to check for clarity and correctness.

**Comment by rhornung67** [comment]
@davidbeckingsale can you double-check the changes in this commit https://github.com/LLNL/RAJA/pull/1119/commits/ca2ca2d1fefb035b634075485feba12ad2056d7b for correctness.

**Comment by rhornung67** [comment]
@davidbeckingsale please look over the config_options.rst file changes in this commit: https://github.com/LLNL/RAJA/pull/1119/commits/d8ed5f9bc8cd93d8423daf0d678aa0fb24fdd945. I think this is much clearer, but still not perfect. Please let me know what you think.

File: docs/sphinx/user_guide/config_options.rst
Comment (kab163) [inline docs/sphinx/user_guide/config_options.rst:28]:
```suggestion
and those that are similar to the standard CMake options or options 
```

Comment (rhornung67) [reply docs/sphinx/user_guide/config_options.rst]: Thanks. Fixed.

**Comment by kab163** [comment]
Do you also happen to have a URL to the RTD page so I can look at the generated page? I can't seem to get to that page from the RTD site - wondering if I have access.

**Comment by rhornung67** [comment]
> Do you also happen to have a URL to the RTD page so I can look at the generated page? I can't seem to get to that page from the RTD site - wondering if I have access.

I enabled the docs on that branch, you should see it on the list of versions on the RAJA RTD project. It's building now. It should be available in a few minutes.

**Comment by kab163** [comment]
> > Do you also happen to have a URL to the RTD page so I can look at the generated page? I can't seem to get to that page from the RTD site - wondering if I have access.
> 
> I enabled the docs on that branch, you should see it on the list of versions on the RAJA RTD project. It's building now. It should be available in a few minutes.

Yes, it works now, thanks Rich!

**Comment by kab163** [comment]
One question: When I go to the config page on RTD here
https://raja.readthedocs.io/en/feature-add-option-prefix/sphinx/user_guide/config_options.html
and go to the blue note, it looks like the note is encompassing the entire table - that's not supposed to happen, is it?

File: docs/sphinx/user_guide/config_options.rst
Comment (davidbeckingsale) [inline docs/sphinx/user_guide/config_options.rst:125]:
```suggestion
      (RAJA_)ENABLE_TESTS          On 
```

Do you think using parentheses for the dependent options might be clearer still? The most confusing thing (that we want to avoid) is that for something like CUDA, you _have_ to say ENABLE_CUDA=On (because RAJA_ENABLE_CUDA=On won't change the BLT option)

Comment (rhornung67) [reply docs/sphinx/user_guide/config_options.rst:125]: That's a good idea. I will work on that change. Thanks.

**Comment by rhornung67** [comment]
> One question: When I go to the config page on RTD here https://raja.readthedocs.io/en/feature-add-option-prefix/sphinx/user_guide/config_options.html and go to the blue note, it looks like the note is encompassing the entire table - that's not supposed to happen, is it? (I think it is messing up some of the formatting for the other text too)

Whoops! I'll fix it. I planned to take a pass through. But thanks for catching that.

File: docs/sphinx/user_guide/config_options.rst
Comment (kab163) [inline docs/sphinx/user_guide/config_options.rst:124]: Shouldn't the parentheses be around "RAJA_" not just "RAJA"?

Comment (rhornung67) [reply docs/sphinx/user_guide/config_options.rst:124]: Grrff... I'm fixing other issues in the docs and am not being careful. Fixes coming.

**Comment by rhornung67** [comment]
@davidbeckingsale @kab163 Finally, I think we are good with the docs. Sorry for the churn. Please take a look and approve if it looks and reads OK.

File: docs/sphinx/user_guide/config_options.rst
Comment (rhornung67) [reply docs/sphinx/user_guide/config_options.rst]: Good catch. Fixed. Docs should be served up in a few minutes.

File: docs/sphinx/user_guide/config_options.rst
Comment (kab163) [inline docs/sphinx/user_guide/config_options.rst:187]:
```suggestion
Turning the ``(RAJA_)ENABLE_CLANG_CUDA`` variable on will build CUDA 
```

Comment (kab163) [reply docs/sphinx/user_guide/config_options.rst:187]: Found one more!

**Review by kab163 — APPROVED** [review]

**Comment by davidbeckingsale** [comment]
Looks good to me. I can't approve this PR since I created it 😄

**Comment by rhornung67** [comment]
@kab163 I fixed a couple more minor issues. Can you re-approve when you have a minute? I'll merge this when checks come back green.

**Review by rhornung67 — APPROVED** [review]

**Comment by rhornung67** [comment]
@kab163 @davidbeckingsale are we good with merging this?

**Pull request merged**
